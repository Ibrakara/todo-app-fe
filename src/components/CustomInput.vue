<template>
  <div class="add-todo-container">
    <input class="todo-input" v-model="newTodo" placeholder="Add a new todo" />
    <button class="add-todo-button" @click="addTodo">Add</button>
  </div>
</template>
<script>
export default {
  name: "CustomInput",
  data() {
    return {
      newTodo: "",
      todos: [],
    };
  },
};
</script>

<style scoped>
.add-todo-container {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 20px;
}
.todo-input {
  padding: 10px;
  font-size: 17px;
  border: none;
  outline: none;
  border-top-left-radius: 17px;
  border-bottom-left-radius: 17px;
  max-width: 500px;
  transition: black 200ms ease-in-out;
  width: 100%;
  height: 40px;
  padding-left: 20px;
  background-color: #1c1f2b;
  color: #fefefe;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
}
.todo-input::placeholder {
  color: #cccccc;
  opacity: 0.8;
}
.add-todo-button {
  background-color: #1dd1a1cc;
  height: 40px;
  border: none;
  border-top-right-radius: 17px;
  border-bottom-right-radius: 17px;
  cursor: pointer;
  width: 50px;
  font-weight: 500;
  font-size: 17px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  transition: background-color 200ms ease, transform 150ms ease;
}
.add-todo-button:hover {
  background-color: #1dd1a1;
  transform: scale(1.05);
}
.add-todo-button:active {
  transform: scale(0.95); /* Slightly shrink the button */
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.3); /* Reduce shadow for a pressed effect */
}
</style>
