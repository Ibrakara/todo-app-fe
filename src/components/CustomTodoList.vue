<template>
  <div class="todo-list">
    <div
      v-for="todo in todos"
      :key="todo._id"
      class="todo-item"
      :class="{ completed: todo.completed }"
    >
      <span
        :class="{ completed: todo.completed }"
        class="todo-text"
        @click="$emit('toggleComplete', todo._id)"
      >
        {{ todo.title }}
      </span>
      <button @click="$emit('deleteTodo', todo._id)" class="delete-button">
        Delete
      </button>
    </div>
  </div>
</template>

<script>
export default {
  name: "CustomTodoList",
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
};
</script>

<style scoped>
.todo-list {
  margin-top: 20px;
  max-width: 500px;
  margin-left: auto;
  margin-right: auto;
}
.todo-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #1c1f2b;
  padding: 10px 20px;
  margin-bottom: 10px;
  border-radius: 10px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  color: #fefefe;
  transition: background-color 200ms ease;
}
.todo-item.completed {
  background-color: #1dd1a1;
  color: #ffffffcc;
}
.todo-text {
  flex-grow: 1;
  text-align: left;
  color: #fefefe;
  cursor: pointer;
}
.todo-text.completed {
  text-decoration: line-through;
}
.delete-button {
  background-color: #ff6b6b;
  border: none;
  border-radius: 5px;
  color: white;
  padding: 5px 10px;
  cursor: pointer;
  transition: background-color 200ms ease, transform 150ms ease;
}
.delete-button:hover {
  background-color: #ff4c4c;
  transform: scale(1.1);
}
.delete-button:active {
  transform: scale(0.95);
}
</style>
