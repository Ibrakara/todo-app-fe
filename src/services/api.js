import axios from "axios";

const apiClient = axios.create({
  baseURL: "http://localhost:3000/api/todos",
  headers: {
    "Content-Type": "application/json",
  },
});

export const getTodos = () => apiClient.get("/");

export const addTodo = (todo) => apiClient.post("/", todo);

export const toggleTodo = (id, completed) =>
  apiClient.put(`/${id}`, { completed });

export const deleteTodo = (id) => apiClient.delete(`/${id}`);
