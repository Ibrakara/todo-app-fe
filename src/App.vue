<template>
  <div id="app">
    <h1 class="app-title">Todo App</h1>
    <CustomInput />
  </div>
</template>

<script>
import CustomInput from "./components/CustomInput.vue";
export default {
  name: "App",
  components: { CustomInput },
  data() {
    return {
      newTodo: "",
      todos: [],
    };
  },
};
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  padding: 2rem 1rem;
  min-height: 100vh;
  background: linear-gradient(130deg, #6a11cb, #2575fc);
  color: #fefefe;
}
.app-title {
  color: #ffffffcc;
  text-shadow: 2px 2px 2px #0008;
}
</style>
