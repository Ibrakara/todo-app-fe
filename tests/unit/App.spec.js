import { mount } from "@vue/test-utils";
import App from "@/App.vue";
import CustomInput from "@/components/CustomInput.vue";
import CustomTodoList from "@/components/CustomTodoList.vue";
import { getTodos, addTodo, toggleTodo, deleteTodo } from "@/services/api.js";

// Mock the API module
jest.mock("@/services/api.js", () => ({
  getTodos: jest.fn(),
  addTodo: jest.fn(),
  toggleTodo: jest.fn(),
  deleteTodo: jest.fn(),
}));

describe("App.vue", () => {
  let wrapper;
  const mockTodos = [
    { _id: "1", title: "Test Todo 1", completed: false },
    { _id: "2", title: "Test Todo 2", completed: true },
  ];

  beforeEach(() => {
    jest.clearAllMocks();

    getTodos.mockResolvedValue({ data: mockTodos });
    addTodo.mockImplementation((todo) =>
      Promise.resolve({ data: { ...todo, _id: "3", completed: false } })
    );
    toggleTodo.mockImplementation((id, completed) =>
      Promise.resolve({ data: { _id: id, completed } })
    );
    deleteTodo.mockResolvedValue({});

    wrapper = mount(App);
  });

  it("renders correctly", () => {
    expect(wrapper.find(".app-title").text()).toBe("Todo App");
    expect(wrapper.findComponent(CustomInput).exists()).toBe(true);
    expect(wrapper.findComponent(CustomTodoList).exists()).toBe(true);
  });

  it("fetches todos on mount", async () => {
    expect(getTodos).toHaveBeenCalledTimes(1);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.todos).toEqual(mockTodos);
  });

  it("adds a new todo", async () => {
    wrapper.vm.newTodo = "New Test Todo";

    await wrapper.vm.addTodo();

    expect(addTodo).toHaveBeenCalledWith({ title: "New Test Todo" });
    expect(wrapper.vm.todos).toHaveLength(3);
    expect(wrapper.vm.newTodo).toBe("");
  });

  it("does not add empty todo", async () => {
    wrapper.vm.newTodo = "   ";
    await wrapper.vm.addTodo();
    expect(addTodo).not.toHaveBeenCalled();
    expect(wrapper.vm.todos).toEqual(mockTodos);
  });

  it("toggles todo completion", async () => {
    await wrapper.vm.toggleComplete("1");
    expect(toggleTodo).toHaveBeenCalledWith("1", true);
    expect(wrapper.vm.todos.find((t) => t._id === "1").completed).toBe(true);
  });

  it("deletes a todo", async () => {
    await wrapper.vm.removeTodo("2");
    expect(deleteTodo).toHaveBeenCalledWith("2");
  });

  it("handles API errors gracefully", async () => {
    getTodos.mockRejectedValueOnce(new Error("Network error"));
    console.error = jest.fn();

    await wrapper.vm.fetchTodos();
    expect(console.error).toHaveBeenCalledWith(
      "Error fetching todos:",
      expect.any(Error)
    );
  });

  it("matches snapshot", async () => {
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
