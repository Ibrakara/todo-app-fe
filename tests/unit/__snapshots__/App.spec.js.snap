// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue matches snapshot 1`] = `
<div id="app">
  <h1 class="app-title">Todo App</h1>
  <div class="add-todo-container"><input placeholder="Add a new todo" class="todo-input"> <button class="add-todo-button">Add</button></div>
  <div class="todo-list">
    <div class="todo-item"><span class="todo-text">
      Test Todo 1
    </span> <button class="delete-button">
        Delete
      </button></div>
    <div class="todo-item completed"><span class="todo-text completed">
      Test Todo 2
    </span> <button class="delete-button">
        Delete
      </button></div>
    <div class="todo-item"><span class="todo-text">
      New Test Todo
    </span> <button class="delete-button">
        Delete
      </button></div>
  </div>
</div>
`;
